### PR TITLE
Add Hyperparameter Optimization (HPO) examples using Ray Tune and HyperNOs

### PR DESCRIPTION
**Pull request by MaxGhi8**
This PR introduces a comprehensive set of examples for hyperparameter optimization (HPO) in DeepXDE, leveraging Ray Tune and the HyperNOs (https://github.com/lu-group/HyperNOs) framework. 

The examples cover the core functionalities of the library:
  - Operator Learning: 1D Poisson, 1D Advection, and 2D Advection (mapped from 1D time-dependent) using PI-DeepONet.
  - PINN Forward Problems: 1D Diffusion architecture optimization.
  - PINN Inverse Problems: Architecture optimization for parameter identification in 1D Diffusion.  

These examples provide a unified, "supervised operator style" template for HPO that is easily extensible to other physics-informed learning tasks.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,28 +12,28 @@
 DeepXDE is a library for scientific machine learning and physics-informed learning. DeepXDE includes the following algorithms:
 
 - physics-informed neural network (PINN)
-    - solving different problems
```
Comment (echen5503) [inline README.md:15]: Don't randomly change white space, keep minimal changes.

**Review by echen5503 — CHANGES_REQUESTED** [review]
Please provide docs for at least one of the examples. Follow the format as in #2059. 

Additionally, the examples themselves seem dubious. Not every hyperparameter needs to be tuned at once, additionally it seems that each example is mostly boilerplate. Please make each example more unique and show different aspects of hyperparam tuning.

**Comment by MaxGhi8** [comment]
Thank you for the feedback. I have refactored the PR as requested:

- Consolidated Example: Replaced redundant scripts with a single master example advection_2d_hpo.py.
- Enhanced HPO: Expanded the search space to include architecture, optimizer dynamics, and activations.
- Documentation: Added a dedicated .rst file in docs/demos/.
- Clean Repo: Reverted unrelated changes to README.md.

**Comment by echen5503** [comment]
Great, this looks much better now. Can you provide screenshots of the documentation and outputs of the code so that we can be sure it runs correctly?

**Comment by MaxGhi8** [comment]
Running with `runs_per_cpu=4.0` on a 12-core machine, the code successfully executed 3 parallel runs simultaneously, as shown in the attached screenshots of the source, start, and final output of the main python function added. I also attached a preview of the documentation. I only made a minor change in the documentation and ran the formatter.

start-code
mid-code
final-code
preview-doc

File: examples/operator/advection_2d_hpo.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
+"""
```
Comment (echen5503) [inline examples/operator/advection_2d_hpo.py]:
```suggestion
"""Backend Supported: pytorch
```

File: examples/operator/advection_2d_hpo.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+Backend supported: pytorch
```
Comment (echen5503) [inline examples/operator/advection_2d_hpo.py]:
```suggestion
```

File: examples/operator/poisson_1d_hpo.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
+"""
```
Comment (echen5503) [inline examples/operator/poisson_1d_hpo.py]: same here. remove the enter

File: README.md [hunk 5]
```diff
@@ -105,6 +106,7 @@ $ git clone https://github.com/lululxvi/deepxde.git
 - [Demos of forward problems](https://deepxde.readthedocs.io/en/latest/demos/pinn_forward.html)
 - [Demos of inverse problems](https://deepxde.readthedocs.io/en/latest/demos/pinn_inverse.html)
 - [Demos of operator learning](https://deepxde.readthedocs.io/en/latest/demos/operator.html)
+- [Demos of hyperparameter optimization](https://deepxde.readthedocs.io/en/latest/demos/operator/advection_2d_hpo.html)
```
Comment (echen5503) [inline README.md:109]: It could be considered that this may not be a very extensible approach. If someone were to add more hyperparam optimization examples, then they would have to individually link them.

Hyperparameter optimization doesn't fit neatly inside of operator, inverse, or forward, maybe you could create a `hyperparameter` folder, where it would also be extensible to people wanting to add examples of learning rate annealing, optuna, etc.

File: docs/demos/operator/advection_2d_hpo.rst [hunk 6]
```diff
@@ -0,0 +1,205 @@
+2D advection: Comprehensive HPO with HyperNOs
+=============================================
+
```
Comment (echen5503) [inline docs/demos/operator/advection_2d_hpo.rst]: Can you put the approximate runtime here? Hyperparam optimization often takes a long time, so the user should be assured of the code's runtime.

**Comment by echen5503** [comment]
In the future, please actually build the documentation and check it, because unexpected issues might arise in the production build. 
image
This time, I did it for you; it looks good.

File: docs/demos/operator/advection_2d_hpo.rst [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+Running the Example
+-------------------
+
+To run this HPO demo, ensure you have ``HyperNOs`` and ``ray[tune]`` installed. You can execute the script directly from the command line:
+
+.. code-block:: bash
+
+    python examples/operator/advection_2d_hpo.py
+
+The script will launch the Ray Tune dashboard where you can monitor the progress of each trial in real-time. Once finished, it will print the best configuration and the corresponding relative loss.
```
Comment (echen5503) [inline docs/demos/operator/advection_2d_hpo.rst]: let the user see the final code here, it is standard for all the other documentation examples.

**Review by echen5503 — CHANGES_REQUESTED** [review]

**Comment by MaxGhi8** [comment]
Hello, I have just pushed the latest updates. I followed your feedback and implemented all the suggested changes. Specifically, I have reorganized the files by moving the examples and documentation into dedicated `hyperparameter` directories. Thank you for the guidance!

File: docs/demos/operator.rst [hunk 8]
```diff
@@ -38,3 +38,4 @@ PI-DeepONet
 -
     - `Diffusion reaction equation with aligned points using ZCS <https://github.com/lululxvi/deepxde/tree/master/examples/operator/diff_rec_aligned_zcs_pideeponet.py>`_
     - `Stokes flow with aligned points using ZCS <https://github.com/lululxvi/deepxde/tree/master/examples/operator/stokes_aligned_zcs_pideeponet.py>`_
+
```
Comment (echen5503) [inline docs/demos/operator.rst]: what's the point of this?

File: docs/demos/hyperparameter/advection_2d_hpo.rst [hunk 9]
```diff
@@ -0,0 +1,459 @@
+2D advection: Comprehensive HPO with HyperNOs
+=============================================
+
+This example demonstrates a rich hyper-parameter optimization (HPO) for a 2D advection problem using `HyperNOs <https://github.com/MaxGhi8/HyperNOs>`_ and `Ray Tune <https://docs.ray.io/en/latest/tune/index.html>`_.
+
+In scientific machine learning, finding the right combination of network architecture, optimizer settings, and non-linear activation functions is often more of an art than a science. This demo showcases how to automate this process by exploring a high-dimensional search space efficiently.
+
+.. note::
+
+    This code takes about 5 minutes to run.
```
Comment (echen5503) [inline docs/demos/hyperparameter/advection_2d_hpo.rst]: mention what computer you ran this on.

File: docs/demos/hyperparameter.rst [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+Here are some demos of hyper-parameter optimization for scientific machine learning problems.
+
+Operator learning
+-----------------
+
+.. toctree::
+   :maxdepth: 1
+
+   hyperparameter/advection_2d_hpo
+
```
Comment (echen5503) [inline docs/demos/hyperparameter.rst]: remove this space, it looks weird

Image

**Review by echen5503 — CHANGES_REQUESTED** [review]
Looking much better. just a few small changes before I think it's good.

**Comment by MaxGhi8** [comment]
I've implemented the requested changes.

**Comment by echen5503** [comment]
Ok. Looks good now.
